File: P/739_Daily_Temperatures.java
//Stack
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        Stack<Integer> temp = new Stack<>();
        Stack<Integer> index = new Stack<>();
        
        int n = temperatures.length;
        int[] answer = new int[n];
        for(int i = 0; i < n; i++) {
            while(!temp.isEmpty() && temp.peek() < temperatures[i]) {
                answer[index.peek()] = i - index.pop();
                temp.pop();
            }
            
            if(i < n - 1 && temperatures[i] < temperatures[i + 1]) {
                answer[i] = 1;
            } else {
                temp.push(temperatures[i]);
                index.push(i);
            }
        }
        return answer;
    }
}

//Stack
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        Stack<Integer> temp = new Stack<>();
        Stack<Integer> index = new Stack<>();
        
        int n = temperatures.length;
        int[] answer = new int[n];
        for(int i = 0; i < n - 1; i++) {
            temp.push(temperatures[i]);
            index.push(i);
            
            while(!temp.isEmpty() && temp.peek() < temperatures[i + 1]) {
                answer[index.peek()] = i + 1 - index.pop();
                temp.pop();
            }
        }
        return answer;
    }
}

//Brutal force(Time: O(n^2) Space: O(n))
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] answer = new int[temperatures.length];
        
        for(int i = 0; i < temperatures.length - 1; i++) {
            int j = i + 1;
            while(j < temperatures.length && temperatures[j] <= temperatures[i]) {
                j++;
            }
            answer[i] = j < temperatures.length ? j - i : 0;
        }
        return answer;
    }
}

File: 주영/stack/739. Daily Temperatures.js
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function (temperatures) {
  let len = temperatures.length;
  let answer = [];
  let stack = [];

  for (let i = 0; i < len; i++) {
    while (stack && temperatures[stack[stack.length - 1]] < temperatures[i]) {
      answer[stack[stack.length - 1]] = i - stack.pop();
    }
    stack.push(i);
    console.log('stack: ', stack);
    console.log('answer: ', answer);
  }
};

let temperatures = [39, 20, 70, 36, 30, 60, 80, 1];
dailyTemperatures(temperatures);

// Stack을 이용한 방법의 시간 복잡도는 대부분 O(N)이다.
